sklearn/manifold/_utils.pyx

```diff
@@ -63,7 +63,7 @@ cpdef cnp.ndarray[cnp.float32_t, ndim=2] _binary_search_perplexity(
 
     # This array is later used as a 32bit array. It has multiple intermediate
     # floating point additions that benefit from the extra precision
-    cdef cnp.ndarray[cnp.float64_t, ndim=2] P = np.zeros(
+    cdef cnp.float64_t[:, :] P = np.zeros(
         (n_samples, n_neighbors), dtype=np.float64)
 
     for i in range(n_samples):
```

Comment:
You need to change the return line by:

```python
return P.astype(np.float32)
```

Reply:
This is not the current behavior. Currently it returns a float64 array (this is because it's using the def part of the function which doesn't care about the return type)

Reply:
I'd rather keep the current behavior